ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,529 @@
+package wal
+
+import (
+	"bufio"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/ledger/complete/mtrie/flattener"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/node"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+	utilsio "github.com/onflow/flow-go/utils/io"
+)
+
+const subtrieLevel = 4
+const subtrieCount = 1 << subtrieLevel
```

Comment:
Do we need to handle the very rare case of trie not having 4 levels and only have very small number of registers? maybe just returning error in that case is good enough

Reply:
This is already handled. In this case, there would be no subtrie nodes, we would still produce part files, but they have no subtrie nodes. The nodes are just top level nodes.